src/coreclr/src/jit/gentree.cpp

```diff
@@ -3295,57 +3291,192 @@ unsigned Compiler::gtSetEvalOrder(GenTree* tree)
+                        //    movn which loads into any halfword zeroing the remaining halfwords then bitwise inverting the register
+                        // In some cases it is preferable to use movn, because it has the side effect of filling the other halfwords
+                        // with ones
+
+                        // Determine whether movn or movz will require the fewest instructions to populate the immediate
+                        bool preferMovz = false;
+                        bool preferMovn = false;
+                        int  instructionCount = 4;
+
+                        for (int i = (size == EA_8BYTE) ? 48 : 16; i >= 0; i -= 16)
+                        {
+                            if (!preferMovn && (uint16_t(imm >> i) == 0x0000))
+                            {
+                                preferMovz = true; // by using a movk to start we can save one instruction
+                                instructionCount--;
+                            }
+                            else if (!preferMovz && (uint16_t(imm >> i) == 0xffff))
+                            {
+                                preferMovn = true; // by using a movn to start we can save one instruction
+                                instructionCount--;
+                            }
+                        }
+
+                        costSz = 4 * instructionCount;
+                        costEx = instructionCount;
+                    }
+                }
                 goto COMMON_CNS;
             }
 
 #elif defined(TARGET_ARM64)
-            case GT_CNS_LNG:
+
             case GT_CNS_STR:
+            case GT_CNS_LNG:
             case GT_CNS_INT:
-                // TODO-ARM64-NYI: Need cost estimates.
-                costSz = 1;
-                costEx = 1;
+                if (JitConfig.JitDisableConstCSE() == 1)
+                {
+                    costSz = 1;
+                    costEx = 1;
+                }
+                else
+                {
+                    GenTreeIntConCommon* con = tree->AsIntConCommon();
+                    bool iconNeedsReloc = con->ImmedValNeedsReloc(this);
+                    INT64 imm = con->LngValue();
+                    emitAttr size = EA_8BYTE;
+
+                    if ((imm >= -256) && (imm < 1024))
+                    {
+                        costSz = 2;
+                        costEx = 1;
+                    }
+                    else if (emitter::emitIns_valid_imm_for_mov(imm, size))
+                    {
+                        costSz = 4;
+                        costEx = 1;
+                    }
+                    else if (iconNeedsReloc)
+                    {
+                        costSz = 8;
+                        costEx = 2;
+                    }
+                    else
+                    {
+                        // Arm64 allows any arbitrary 16-bit constant to be loaded into a register halfword
```

Comment:
HWIntrinsics recently got some support in lowering where we will generate better code if all inputs to a `NI_Vector128_Create` call are constant.
Is there any handling that should be added here to help account for that?

Reply:
No because the Lower phase runs after the CSE phase